CASSANDRA-17198 Support for LIKE  expressions as PostFilter in SAI

pranavshenoy opened the pull request:
[CASSANDRA-17198](https://issues.apache.org/jira/browse/CASSANDRA-17198) Support for LIKE expression as PostFilter in SAI

Since LIKE is filtered in post filter, ALLOW FILTERING is required. I tested it with some basic queries. Will do further manual testing + unit tests. It would be helpful for early reviews. 


patch by Pranav Shenoy; reviewed by Caleb Rackliffe for CASSANDRA-17198

maedhroz commented on src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java:
```diff
@@ -206,11 +206,20 @@ public StatementRestrictions(ClientState state,
+                    if (!restriction.hasSupportingIndex(indexRegistry))
+                        throw invalidRequest("%s restriction is only supported on properly " +
+                                                        "indexed columns. %s is not valid.", operator, relation);
+                }
```
I'm guessing this has been changed already in your not-yet-pushed work, but once LIKE doesn't return `true` from `requiresIndexing()`, seems like this change should no longer be necessary.

pranavshenoy replied:
yeah. Have reverted this. Thanks

maedhroz commented on src/java/org/apache/cassandra/utils/ByteBufferUtil.java:
```diff
@@ -955,6 +955,18 @@ public static boolean contains(ByteBuffer buffer, ByteBuffer subBuffer)
+    public static boolean matches(ByteBuffer buffer1, ByteBuffer buffer2)
+    {
+        return buffer1.equals(buffer2);
+    }
```
nit: I'd probably just inline this in `Expression` rather than having a one-line utility method.

pranavshenoy replied:
done. Thanks

maedhroz commented on src/java/org/apache/cassandra/index/sai/plan/Expression.java:
```diff
@@ -268,6 +287,10 @@ public boolean isSatisfiedBy(ByteBuffer columnValue)
         {
+            // Post-filtering for LIKE operators must be consistent with implementation of LIKE operators in non-SAI Allow Filtering path
+            if(operator.isLikeVariant())
+                return termMatches(value.raw, lower.value.raw);
```
I think we still may need to account for things like case-insensitive search eventually, so we should be able to just fall through to `validateStringValue()`, which would just call `termMatches()` anyway?

pranavshenoy committed:
CASSANDRA-17198 Support for LIKE expression as PostFilter in SAI

    Since LIKE is filtered in post filter, ALLOW FILTERING is required.
    I tested it with some basic queries. Will do further manual testing + unit tests.
    It would be helpful for early reviews.

    patch by Pranav Shenoy; reviewed by Caleb Rackliffe for CASSANDRA-17198

maedhroz commented on test/unit/org/apache/cassandra/index/sai/cql/UnindexedExpressionsTest.java:
```diff
@@ -63,7 +63,7 @@ public void unsupportedOperatorsAreHandledTest() throws Throwable
 
         // The LIKE operator is rejected because it needs to be handled by an index
-        assertInvalidMessage("LIKE restriction is only supported on properly indexed columns",
+        assertInvalidMessage("Cannot execute this query as it might involve data filtering and thus may have unpredictable performance. If you want to execute this query despite the performance unpredictability, use ALLOW FILTERING",
```
nit: Can we just use `StatementRestrictions.REQUIRES_ALLOW_FILTERING_MESSAGE` here?

pranavshenoy replied:
done

maedhroz commented on test/unit/org/apache/cassandra/index/sai/cql/AllowFilteringTest.java:
```diff
@@ -433,4 +433,83 @@ public void testAllowFilteringDuringIndexBuild() throws Throwable
+        assertRowCount(execute("SELECT * FROM %s WHERE k3 > 0 AND k2 LIKE 'f%%' ALLOW FILTERING"), 3);
+        assertRowCount(execute("SELECT * FROM %s WHERE k3 > 0 AND k2 LIKE 'ca%%' ALLOW FILTERING"), 0);
+        assertThatThrownBy(() -> execute("SELECT * FROM %s WHERE k3 > 0 AND k2 LIKE 'f%%'"))
+                .hasMessageContaining("use ALLOW FILTERING")
```
nit: Can we use `StatementRestrictions.REQUIRES_ALLOW_FILTERING_MESSAGE`?

maedhroz replied:
...and in the tests below, etc.

pranavshenoy replied:
done

maedhroz commented on test/unit/org/apache/cassandra/cql3/validation/entities/SecondaryIndexTest.java:
```diff
@@ -843,16 +844,16 @@ public void prepareStatementsWithLIKEClauses() throws Throwable
         // LIKE is not supported on indexes of non-literal values
         // this is rejected before binding, so the value isn't available in the error message
-        assertInvalidMessage("LIKE restriction is only supported on properly indexed columns. v3 LIKE ? is not valid",
+        assertInvalidMessage("Cannot execute this query as it might involve data filtering and thus may have unpredictable performance. If you want to execute this query despite the performance unpredictability, use ALLOW FILTERING",
```
nit: Can we use `StatementRestrictions.REQUIRES_ALLOW_FILTERING_MESSAGE`?

pranavshenoy replied:
done

maedhroz commented on test/unit/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictionsTest.java:
```diff
@@ -1669,7 +1669,7 @@ private static Restriction newMultiEq(TableMetadata tableMetadata, int firstInde
         return new SimpleRestriction(ColumnsExpression.multiColumns(columns, tableMetadata),
                                      Operator.EQ,
-                                     Terms.of(new MultiElements.Value(tupleType, asList(values))));
+                                     Terms.of(new MultiElements.Value(tupleType, asList(values))), false);
```
nit/formatting: I'd move these `false` arguments to the next line so they're on the same level of indentation as the other arguments.

pranavshenoy replied:
done

maedhroz commented on src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java:
```diff
@@ -345,11 +352,14 @@ public void addToRowFilter(RowFilter filter, IndexRegistry indexRegistry, QueryO
+                    {
+                        throw invalidRequest("%s is only supported on properly indexed columns or with ALLOW FILTERING",
+                                            expression);
+                    }
```
nit: I might just go single-line `if` here. It probably won't ever be multi-line...

```
if (!index.isPresent() && !allowFiltering)
    throw invalidRequest("%s is only supported on properly indexed columns or with ALLOW FILTERING", expression);
```

pranavshenoy replied:
done

maedhroz commented on src/java/org/apache/cassandra/cql3/Ordering.java:
```diff
@@ -105,7 +105,7 @@ public SingleRestriction toRestriction()
             return new SimpleRestriction(ColumnsExpression.singleColumn(columnMetadata, tableMetadata),
                                          Operator.ANN,
-                                         Terms.of(vectorValue));
+                                         Terms.of(vectorValue), false);
```
nit/formatting: I'd bump `false` to the next line so the arguments are all at the same indentation level.

pranavshenoy replied:
done

maedhroz reviewed (approved):
Approved w/ minor nits. Trying to get a clean CI run...

maedhroz commented:
Committed as https://github.com/apache/cassandra/commit/cf806cac1aca7b56fe715bbd2641cda65beef91e
